aconcli/service/aconclient_http.go

```diff
@@ -0,0 +1,465 @@
+	if err = json.Unmarshal(content, &r); err != nil {
+		return "", nil, fmt.Errorf(clientUnmarshalErrFmt, "AddManifest", err)
+	}
+	return r.ImageId, r.MissingLayers, nil
+}
+
+func (c *AconClientHttp) AddBlob(alg uint32, blobpath string) error {
+	blobpath = filepath.Clean(blobpath)
+	requestURL := fmt.Sprintf("%s/%s?%s=%s", c.makeURL(endpointBlob),
+		filepath.Base(blobpath), fieldAlg, strconv.FormatUint(uint64(alg), 10))
+	body := &bytes.Buffer{}
+	w := multipart.NewWriter(body)
+	if err := multipartFile(w, fieldBlob, blobpath); err != nil {
+		return fmt.Errorf("AddBlob, prepare multipart error: %s", err)
+	}
+	w.Close()
+
+	req, err := http.NewRequest(http.MethodPut, requestURL, body)
+	if err != nil {
+		return fmt.Errorf(clientMakeReqErrFmt, "AddBlob", err)
+	}
+	req.Header.Add("Content-Type", w.FormDataContentType())
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return fmt.Errorf(clientSendReqErrFmt, "AddBlob", err)
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "AddBlob", err)
+	} else {
+		return nil
+	}
+}
+
+func (c *AconClientHttp) Finalize() error {
+	requestURL := c.makeURL(endpointFinalize)
+	req, err := http.NewRequest(http.MethodPost, requestURL, nil)
+	if err != nil {
+		return fmt.Errorf(clientMakeReqErrFmt, "Finalize", err)
+	}
+
+	resp, err := c.client.Do(req)
+	if err != nil {
+		return fmt.Errorf(clientSendReqErrFmt, "Finalize", err)
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "Finalize", err)
+	} else {
+		return nil
+	}
+}
+
+func (c *AconClientHttp) Start(imageId string, env []string) (uint32, error) {
+	requestURL := c.makeURL(endpointStart)
+	d := url.Values{}
+	d.Set(fieldImgeId, imageId)
+	for _, e := range env {
+		d.Add(fieldEnvs, e)
+	}
+
+	resp, err := c.client.PostForm(requestURL, d)
+	if err != nil {
+		return 0, fmt.Errorf(clientSendReqErrFmt, "Start", err)
+	}
+	content, err := processReponse(resp)
+	if err != nil {
+		return 0, fmt.Errorf(clientProcRespErrFmt, "Start", err)
+	}
+	r := StartResponse{}
+	if err = json.Unmarshal(content, &r); err != nil {
+		return 0, fmt.Errorf(clientUnmarshalErrFmt, "Start", err)
+	}
+	return r.ContainerId, nil
+}
+
+func (c *AconClientHttp) Kill(cid uint32, signum int32) error {
+	requestURL := c.makeURL(endpointKill)
+	d := url.Values{}
+	d.Set(fieldConId, strconv.FormatUint(uint64(cid), 10))
+	d.Set(fieldSignum, strconv.FormatInt(int64(signum), 10))
+
+	resp, err := c.client.PostForm(requestURL, d)
+	if err != nil {
+		return fmt.Errorf(clientSendReqErrFmt, "Kill", err)
+	}
+	if _, err = processReponse(resp); err != nil {
+		return fmt.Errorf(clientProcRespErrFmt, "Kill", err)
+	} else {
+		return nil
+	}
+}
+
+func (c *AconClientHttp) Restart(cid uint32, timeout uint64) error {
+	requestURL := c.makeURL(endpointRestart)
+	d := url.Values{}
+	d.Set(fieldConId, strconv.FormatUint(uint64(cid), 10))
+	d.Set(fieldTimeout, strconv.FormatUint(timeout, 10))
```

Comment:
Is there a HTTP timeout? What if the restart timeout is longer than the HTTP timeout? Do we need to set HTTP timeout to be at least as long as restart timeout?

Reply:
Fixed